src/btree/bt_curnext.c

```diff
@@ -245,56 +246,84 @@ __cursor_var_next(WT_CURSOR_BTREE *cbt, bool newpage, bool restart, size_t *skip
+            cbt->iface.value.size = cbt->tmp->size;
+            return (0);
+        }
 
-                /* Adjust for the outer loop increment. */
-                --cbt->recno;
+        /* Unpack the cell and build the return information. */
+        cell = WT_COL_PTR(page, cip);
+        __wt_cell_unpack_kv(session, page->dsk, cell, &unpack);
+        rle = __wt_cell_rle(&unpack);
+        if (unpack.type == WT_CELL_DEL) {
+            if (rle == 1) {
                 ++*skippedp;
                 continue;
             }
 
-            WT_RET(__wt_bt_col_var_cursor_walk_txn_read(session, cbt, page, &unpack, cip));
-            if (cbt->upd_value->type == WT_UPDATE_INVALID ||
-              cbt->upd_value->type == WT_UPDATE_TOMBSTONE) {
-                ++*skippedp;
-                continue;
-            }
+            /*
+             * There can be huge gaps in the variable-length column-store name space appearing as
+             * deleted records. If more than one deleted record, do the work of finding the next
+             * record to return instead of looping through the records.
+             *
+             * First, find the smallest record in the update list that's larger than the current
+             * record.
+             */
+            ins = __col_insert_search_gt(cbt->ins_head, cbt->recno);
 
-            return (0);
+            /*
+             * Second, for records with RLEs greater than 1, the above call to __col_var_search
+             * located this record in the page's list of repeating records, and returned the
+             * starting record. The starting record plus the RLE is the record to which we could
+             * skip, if there was no smaller record in the update list.
+             */
+            cbt->recno = rle_start + rle;
+            if (ins != NULL && WT_INSERT_RECNO(ins) < cbt->recno)
+                cbt->recno = WT_INSERT_RECNO(ins);
+
+            /* Adjust for the outer loop increment. */
+            --cbt->recno;
+            ++*skippedp;
+            continue;
+        }
+
+        WT_RET(__wt_txn_read(session, cbt, NULL, cbt->recno, cbt->ins ? cbt->ins->upd : NULL));
+        if (cbt->upd_value->type == WT_UPDATE_INVALID ||
+          cbt->upd_value->type == WT_UPDATE_TOMBSTONE) {
+            ++*skippedp;
+            continue;
         }
-        cbt->iface.value.data = cbt->tmp->data;
-        cbt->iface.value.size = cbt->tmp->size;
+        WT_RET(__wt_value_return(cbt, cbt->upd_value));
+
+        /*
+         * It is only safe to cache the value for other keys in the same RLE cell if it is globally
+         * visible. Otherwise, there might be some older timestamp where the value isn't uniform
+         * across the cell. Always set cip_saved so it's easy to tell when we change cells.
+         */
+        cbt->cip_saved = cip;
+        if (rle > 1 &&
+          __wt_txn_visible_all(session, unpack.tw.start_txn, unpack.tw.durable_start_ts)) {
+            /*
```

Comment:
Can we compare the txnid and timestamp of the unpack and cbt->upd_value to tell if we are pointing to the disk image or not? I think that should be possible.

Reply:
I think so, but it's not sufficient -- we need to check that all the keys in the RLE cell are globally visible, not just the current one.

Reply:
I'm a bit confused. What we are doing here is to cache an rle value to see if that can be used later for another cell, isn't it? What I suggest is that if cursor is pointing to a value on the disk image, we can cache it. If it is pointing to a value on the update chain, we don't cache it. I don't understand for caching why we need to check all the keys.

Reply:
Because the purpose is to cache it for use with the next key -- therefore, the value also needs to be the correct value for the next key as well as the current one.